src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,478 @@
+                // of the copmilation. Only the MVID, name and version contribute here hence the file path
+                // is deliberately omitted here.
+                if (moduleMetadata.GetMetadataReader() is { IsAssembly: true } peReader)
+                {
+                    var assemblyDef = peReader.GetAssemblyDefinition();
+                    writer.Write("name", peReader.GetString(assemblyDef.Name));
+                    writer.Write("version", assemblyDef.Version.ToString());
+                    WriteByteArrayValue(writer, "publicKey", peReader.GetBlobBytes(assemblyDef.PublicKey).AsSpan());
+                }
+
+                writer.Write("mvid", moduleMetadata.GetModuleVersionId().ToString());
+                writer.WriteKey("properties");
+                writeMetadataReferenceProperties(writer, reference.Properties);
+            }
+            else
+            {
+                // TODO: handle the other reference kinds
+                throw new InvalidOperationException();
+            }
+            writer.WriteObjectEnd();
+
+            static void writeMetadataReferenceProperties(JsonWriter writer, MetadataReferenceProperties properties)
+            {
+                writer.WriteObjectStart();
+                writer.Write("kind", properties.Kind);
+                writer.Write("embedInteropTypes", properties.EmbedInteropTypes);
+                if (properties.Aliases is { Length: > 0 } aliases)
+                {
+                    writer.WriteKey("aliases");
+                    writer.WriteArrayStart();
+                    foreach (var alias in aliases)
+                    {
+                        writer.Write(alias);
+                    }
+                    writer.WriteArrayEnd();
+                }
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private void WriteEmitOptions(JsonWriter writer, EmitOptions? options)
+        {
+            writer.WriteObjectStart();
+            if (options is null)
+            {
+                writer.WriteObjectEnd();
+                return;
+            }
+
+            writer.Write("emitMetadataOnly", options.EmitMetadataOnly);
+            writer.Write("tolerateErrors", options.TolerateErrors);
+            writer.Write("includePrivateMembers", options.IncludePrivateMembers);
+            if (options.InstrumentationKinds.Length > 0)
+            {
+                writer.WriteArrayStart();
+                foreach (var kind in options.InstrumentationKinds)
+                {
+                    writer.Write(kind);
+                }
+                writer.WriteArrayEnd();
+            }
+
+            writeSubsystemVersion(writer, options.SubsystemVersion);
+            writer.Write("fileAlignment", options.FileAlignment);
+            writer.Write("highEntropyVirtualAddressSpace", options.HighEntropyVirtualAddressSpace);
+            writer.Write("baseAddress", options.BaseAddress.ToString());
```

Comment:
this likely needs passing in invariant culture here.  we should likely do an audit of every ToString to make sure it is valid.

Reply:
Moved to `"G"` which should round trip across cultures. I'm verifying that now.

Reply:
Nope I was wrong about this.